cudapoa/src/cudapoa_batch.cpp

```diff
@@ -312,6 +312,83 @@ StatusType CudapoaBatch::get_msa(std::vector<std::vector<std::string>>& msa, std
     return StatusType::success;
 }
 
+void CudapoaBatch::get_graphs(std::vector<DirectedGraph>& graphs, std::vector<StatusType>& output_status)
+{
+    int32_t max_nodes_per_window_ = banded_alignment_ ? CUDAPOA_MAX_NODES_PER_WINDOW_BANDED : CUDAPOA_MAX_NODES_PER_WINDOW;
+    CGA_CU_CHECK_ERR(cudaMemcpyAsync(graph_details_h_->nodes,
+                                     graph_details_d_->nodes,
+                                     sizeof(uint8_t) * max_nodes_per_window_ * max_poas_,
+                                     cudaMemcpyDeviceToHost,
+                                     stream_));
+
+    CGA_CU_CHECK_ERR(cudaMemcpyAsync(graph_details_h_->incoming_edges,
+                                     graph_details_d_->incoming_edges,
+                                     sizeof(uint16_t) * max_nodes_per_window_ * CUDAPOA_MAX_NODE_EDGES * max_poas_,
+                                     cudaMemcpyDeviceToHost,
+                                     stream_));
+
+    CGA_CU_CHECK_ERR(cudaMemcpyAsync(graph_details_h_->incoming_edge_weights,
+                                     graph_details_d_->incoming_edge_weights,
+                                     sizeof(uint16_t) * max_nodes_per_window_ * CUDAPOA_MAX_NODE_EDGES * max_poas_,
+                                     cudaMemcpyDeviceToHost,
+                                     stream_));
+
+    CGA_CU_CHECK_ERR(cudaMemcpyAsync(graph_details_h_->incoming_edge_count,
+                                     graph_details_d_->incoming_edge_count,
+                                     sizeof(uint16_t) * max_nodes_per_window_ * max_poas_,
+                                     cudaMemcpyDeviceToHost,
+                                     stream_));
+
+    CGA_CU_CHECK_ERR(cudaMemcpyAsync(input_details_h_->sequence_lengths,
```

Comment:
not 100% sure, but this might be a bug: do you intend to copy `global_sequence_idx_` many sequence lengths here or just one at `input_details_d_->sequence_lengths+global_sequence_idx_`?

Reply:
or should that be `poa_count_`?

Reply:
ok, I think I get it now, but please double check.

Reply:
The lengths of the updated graphs are stored in the 0th position of the sequence lengths. When optimizing it seemed the best way to re-use memory without having to introduce yet another buffer. And the sequence lengths are a contiguous array with lengths. So we need the 0th index for each POA within this contiguous buffer (that logic is within the loop).

Reply:
so you need the 0th and the `global_sequence_idx_-1`th entry?

Reply:
The global sequence index spans _all_ the sequences passed to device, across all POAs. So for example, if POA 1 had 10 sequences, POA 2 had 15 sequences, POA 3 had 5 sequences, `global_sequence_idx` would be 30. And the indices we would need are 0 (first length for POA 1), 15 (first length for POA 2) and 25 (first length for POA 3)

Reply:
ok, got it. Thanks for the explanation.